src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs

```diff
@@ -0,0 +1,472 @@
+        where T : ICopyable<T>
+    {
+        enum TheAction
+        {
+            None,
+            Move,
+            Add,
+            Insert
+        }
+
+        CompositeDisposable disposables = new CompositeDisposable();
+        IObservable<Unit> source;
+        Func<T, T, int> comparer;
+        Func<T, int, IList<T>, bool> filter;
+        IScheduler scheduler;
+        List<T> original;
+
+        public TrackingCollection(Func<T, T, int> comparer = null, Func<T, int, IList<T>, bool> filter = null, IScheduler scheduler = null)
+        {
+            this.scheduler = scheduler ?? RxApp.MainThreadScheduler;
+            this.comparer = comparer;
+            this.filter = filter;
+            if (filter != null)
+                original = new List<T>();
+        }
+
+        public TrackingCollection(IObservable<T> source, Func<T, T, int> comparer = null, Func<T, int, IList<T>, bool> filter = null, IScheduler scheduler = null)
+            : this(comparer, filter, scheduler)
+        {
+            Listen(source);
+        }
+
+        public ITrackingCollection<T> Listen(IObservable<T> obs)
+        {
+            source = obs
+                .ObserveOn(filter != null ? TaskPoolScheduler.Default : scheduler)
+                .Select(t =>
+                {
+                    var list = filter == null ? this : original as IList<T>;
+
+                    var idx = list.IndexOf(t);
+                    if (idx >= 0)
+                    {
+                        var old = list[idx];
+                        var comparison = 0;
+                        if (comparer != null)
+                            comparison = comparer(t, old);
+                        list[idx].CopyFrom(t);
+
+                        if (comparer == null || comparison == 0)
+                            return Tuple.Create(TheAction.None, idx, idx); // no sorting to be done, just replacing the element in-place
+                        else
+                        {
+                            var i = idx;
+                            if (comparison < 0) // replacing element has lower sorting order, find the correct spot towards the beginning
+                                for (var pos = i - 1; i > 0 && comparer(list[i], list[pos]) < 0; i--, pos--)
+                                    Swap(list, i, pos);
+                            else if (comparison > 0) // replacing element has higher sorting order, find the correct spot towards the end
+                                for (var pos = i + 1; i < list.Count - 1 && comparer(list[i], list[pos]) > 0; i++, pos++)
+                                    Swap(list, i, pos);
+
+                            if (filter == null)
+                            {
+                                OnPropertyChanged(new PropertyChangedEventArgs("Item[]"));
+                                OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Move, old, i, idx));
```

Comment:
I don't understand why we don't change the property event args if filter is null. Is that because a filter causes us to change another list as opposed to `this` list?